tools/trainerproc/main.c

```diff
@@ -0,0 +1,1965 @@
+            parsed->trainers = trainers_;
+        }
+        struct Trainer *trainer = &parsed->trainers[parsed->trainers_n];
+        while (match_empty_line(p)) {}
+        if (match_eof(p))
+            break;
+        if (!parse_trainer(p, parsed, trainer))
+        {
+            struct Token t;
+            // Skip to the next trainer.
+            for (;;)
+            {
+                struct Parser p_ = *p;
+                if (match_eof(&p_) || parse_section(&p_, &t))
+                    break;
+                skip_line(p);
+            }
+        }
+        else
+        {
+            parsed->trainers_n++;
+        }
+    }
+}
+
+static void fprint_string(FILE *f, struct String s)
+{
+    fprintf(f, "%.*s", s.string_n, s.string);
+}
+
+static void fprint_bool(FILE *f, bool b)
+{
+    fprintf(f, b ? "TRUE" : "FALSE");
+}
+
+static void fprint_stats(FILE *f, const char *macro, struct Stats stats)
+{
+    fprintf(f, "%s(%d, %d, %d, %d, %d, %d)", macro, stats.hp, stats.attack, stats.defense, stats.speed, stats.special_attack, stats.special_defense);
+}
+
+static bool is_constant(struct String s, const char *prefix)
+{
+    for (int i = 0;; i++)
+    {
+        if (i == s.string_n)
+            return false;
+        else if (prefix[i] == '\0')
+            return s.string[i] == '_';
+        else if (prefix[i] != s.string[i])
+            return false;
+    }
+}
+
+static void fprint_constant(FILE *f, const char *prefix, struct String s)
+{
+    if (!is_constant(s, prefix)) fprintf(f, "%s_", prefix);
+    if (s.string_n > 0)
+    {
+        for (int i = 0; i < s.string_n; i++)
+        {
+            unsigned char c = s.string[i];
+            if (('A' <= c && c <= 'Z') || ('0' <= c && c <= '9'))
+                fputc(c, f);
+            else if ('a' <= c && c <= 'z')
+                fputc(c - 'a' + 'A', f);
+            else
+                fputc('_', f);
+        }
+    }
+    else
+    {
+        fprintf(f, "NONE");
+    }
+}
+
+// This is a really stupid helper for 'fprint_species'.
+static bool is_utf8_character(struct String s, int *i, const unsigned char *utf8)
+{
+    for (int j = *i;; j++, utf8++)
+    {
+        if (*utf8 == '\0')
+        {
+            *i = j - 1; // '- 1' because of the '++' in 'fprint_species'.
+            return true;
+        }
+        else if (j == s.string_n || s.string[j] != *utf8)
+        {
+            return false;
+        }
+    }
+}
+
+static const struct {
+    const char *literal;
+    const char *constant;
+} literal_species[] =
+{
+    { "Calyrex-Ice", "CALYREX_ICE_RIDER" },
+    { "Calyrex-Shadow", "CALYREX_SHADOW_RIDER" },
+    { "Indeedee-F", "INDEEDEE_FEMALE" },
+    { "Meowstic-F", "MEOWSTIC_FEMALE" },
+    { "Maushold-Four", "MOUSHOLD_FAMILY_OF_FOUR" },
+    { "Tauros-Paldea-Aqua", "TAUROS_PALDEAN_AQUA_BREED" },
+    { "Tauros-Paldea-Blaze", "SPECIES_TAUROS_PALDEAN_BLAZE_BREED" },
+    { "Tauros-Paldea-Combat", "SPECIES_TAUROS_PALDEAN_COMBAT_BREED" },
```

Comment:
Why do these two have the `SPECIES_` prefix but not the rest?

Reply:
Because I made a mistake :sweat_smile: Without `SPECIES_` is correct!